### ISSUE TITLE
Log names of new ctypes

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#165]
I'm finding it surprisingly hard to log the name of each new ctype when it is created. The purpose is to log a mapping from ctype-id to a useful name e.g. to put in IR listings to clarify the types involved in ctype allocations and guards.

Can anyone show how to do this e.g. with a printf?

The solution I have on the `auditlog` branch is incomplete and in practice seems to miss all of the important types: https://github.com/lukego/raptorjit/commit/d0191fec27b49c383a5dd181d5d73cb081c4933f#diff-3ffc1aaef5b69e255ffad67493361e9d.

**Comment by fsfod** [comment]
Couldn't you just do it on demand right before you write out a trace you write out all new all ctypes since you last checked.

**Comment by lukego** [comment]
Good idea! I had thought that polling wouldn't be suitable but yes, logging new types right before a trace should be fine.

**Comment by lukego** [comment]
Hm. That patch is not stable. I'm occasionally seeing segfaults when calling `lj_ctype_repr()`. Have to see if that's a bonehead mistake of mine or some invariant that I don't understand e.g. not being supposed to call this with certain ctypes.

**Comment by lukego** [comment]
Fixed in lukego/raptorjit@473cb199e7741e88f210b904fecccf0129b4da59.

Certain ctype IDs are not printable but will be attempted by the polling loop anyway. These would cause an assertion failure in the VM or else an unpredictable crash if assertions aren't enabled. Now it sets the `ok` flag to false which should simply print them as `"?"`.

I believe I have actually gone down this path and hit the same problem in the past and that was why I didn't poll the types from the beginning.
